File: backend/internal/core/domain/userproduct.go
package domain

import (
	"database/sql"
	"time"

	"github.com/shopspring/decimal"
)

// internal/core/domain/userproduct_query.go

type SortDir string

const (
	SortAsc  SortDir = "asc"
	SortDesc SortDir = "desc"
)

type UserProductQuery struct {
	ShopID int64 `json:"shopId"`

	BrandIDs      []int64 `json:"brandIds"`      // فیلتر براساس چند برند
	CategoryID    int64   `json:"categoryId"`    // دستهٔ اصلی (pb.category_id)
	SubCategoryID int64   `json:"subCategoryId"` // اگر در اسکیمای شما روی product هست: p.sub_category_id

	IsDollar *bool `json:"isDollar"`

	Search string `json:"search"`

	SortUpdated SortDir `json:"sortUpdated"` // asc|desc (پیش‌فرض desc)

	Limit  int `json:"limit"`
	Offset int `json:"offset"`
}

type UserProductView struct {
	UserProduct
	ProductCategory string                          `json:"productCategory"`
	ProductBrand    string                          `json:"productBrand"`
	ProductModel    string                          `json:"productModel"`
	Description     string                          `json:"description"`
	DefaultImageUrl string                          `json:"defaultImageUrl"`
	DefaultFilter   *ProductFilterRelationViewModel `gorm:"-" json:"defaultFilter"`
	IsLiked         bool                            `json:"isLiked"`
	ShopsCount      int32                           `json:"shopsCount"`
}

type ShopViewModel struct {
	ShopInfo *User              `json:"shopInfo"`
	Products []*UserProductView `json:"products"`
}
type SearchProductsData struct {
	ProductItems []*SearchProductViewModel `json:"productItems"`
	Filters      []*ProductFilterData      `json:"filters"`
	Brands       []*ProductBrand           `json:"brands"`
	Models       []*ProductModel           `json:"models"`
	TotalCount   int64                     `json:"totalCount"` // <--- این فیلد اضافه شد
}

type UserProductFilter struct {
	AllowedProductIDs []int64
	CategoryID        int64
	SearchText        string
	Limit             int
	Offset            int
	BrandIDs          int64
	ModelIDs          int64
	SortOrder         SortOrder
}

type SearchProductViewModel struct {
	Product
	CategoryTitle           string          `json:"categoryTitle"`
	BrandTitle              string          `json:"brandTitle"`
	ModelTitle              string          `json:"modelTitle"`
	IsLiked                 bool            `json:"isLiked"`
	Tags                    []*ProductTag   `gorm:"-" json:"tags"`
	FilterOptionIds         []int64         `gorm:"-" json:"filterOptionIds"`
	Price                   decimal.Decimal `gorm:"-" json:"price"`
	LastPriceUpdateDateTime time.Time       `json:"lastPriceUpdateDateTime"`
}

type ShopProduct struct {
	UserProduct
	DefaultImageUrl string `json:"defaultImageUrl"`
	ShopCity        string `json:"shopCity"`
	ShopPhone1      string `json:"shopPhone1"`
	ShopPhone2      string `json:"shopPhone2"`
	ShopPhone3      string `json:"shopPhone3"`
	ShopName        string `json:"shopName"`
	IsLiked         bool   `json:"isLiked"`
	LikesCount      int64  `json:"likesCount"`
}

type ProductShop struct {
	ID              int64           `json:"id"`
	UserID          int64           `json:"userId"`
	FinalPrice      decimal.Decimal `json:"finalPrice"`
	DefaultImageUrl string          `json:"defaultImageUrl"`
	ShopCity        string          `json:"shopCity"`
	ShopPhone1      string          `json:"shopPhone1"`
	ShopPhone2      string          `json:"shopPhone2"`
	ShopPhone3      string          `json:"shopPhone3"`
	ShopName        string          `json:"shopName"`
	IsLiked         bool            `json:"isLiked"`
	LikesCount      int64           `json:"likesCount"`
	UpdatedAt       time.Time       `json:"updatedAt"`
}

type ProductInfoViewModel struct {
	ShopProducts []*ProductShop    `json:"shopProducts"`
	ProductInfo  *ProductViewModel `json:"productInfo"`
}

type ShopsProductViewModel struct {
	ShopProducts []*ShopProduct    `json:"shopProducts"`
	ProductInfo  *ProductViewModel `json:"productInfo"`
}

type ProductPrice struct {
	ProductID int64
	MinPrice  decimal.Decimal
}

type UserProduct struct {
	ID        int64 `json:"id"`
	UserID    int64 `json:"userId" gorm:"not null;index:idx_user_product_unique,unique,priority:1"`
	ProductID int64 `json:"productId" gorm:"not null;index:idx_user_product_unique,unique,priority:2"`

	BrandID    int64  `json:"brandId"     gorm:"->"`
	CategoryID int64  `json:"categoryId"  gorm:"->"`
	ModelName  string `json:"modelName"   gorm:"->;column:model_name"`

	IsDollar    bool                `json:"isDollar"`
	DollarPrice decimal.NullDecimal `json:"dollarPrice"`
	OtherCosts  decimal.NullDecimal `json:"otherCosts"`
	FinalPrice  decimal.Decimal     `json:"finalPrice"`

	Order     int64        `json:"order"   gorm:"column:order_c"`
	IsHidden  bool         `json:"isHidden"`
	CreatedAt time.Time    `json:"createdAt"`
	UpdatedAt sql.NullTime `json:"updatedAt"`
}

func (UserProduct) TableName() string {
	return "user_product"
}

File: frontend/app/types/userproduct/userProduct.ts
// app/types/userproduct/userProduct.ts

// برای decimal ها در فرانت‌اند رشته در نظر می‌گیریم
export type DecimalString = string;

// --------- مدل‌های اصلی سمت سرور (نمایش/ویو) ---------

export type ProductFilterRelationViewModel = {
  id: number;
  filterId: number;
  optionId: number;
};

export type UserProduct = {
  id: number;
  userId: number;
  productId: number;

  brandId: number;     // gorm:"->"
  categoryId: number;  // gorm:"->"
  modelName: string;   // gorm:"->" column:model_name

  isDollar: boolean;
  dollarPrice?: DecimalString | null; // nullable
  otherCosts?: DecimalString | null;  // nullable
  finalPrice: DecimalString;          // required

  order: number;       // column: order_c
  isHidden: boolean;
  createdAt: string;   // ISO
  updatedAt?: string | null; // ISO or null
};

export type UserProductView = UserProduct & {
  productCategory: string;
  productBrand: string;
  productModel: string;
  description: string;
  defaultImageUrl: string;
  defaultFilter?: ProductFilterRelationViewModel | null;
  isFavorite: boolean;
  shopsCount: number;
};

export type ShopViewModel = {
  shopInfo: {
    id: number;
    userName?: string;
    city?: string;
    phoneNumber?: string;
    imageUrl?: string;
    latitude?: number;
    longitude?: number;
    likesCount?: number;
    shopAddress?: string;
    shopPhone1?: string;
    shopPhone2?: string;
    shopName?: string;
    createdAt?: string;
    instagramUrl?: string;
    telegramUrl?: string;
    whatsappUrl?: string;
    websiteUrl?: string;
    updatedAt?: string;

    // سایر فیلدهای کاربر که نیاز داری به تدریج اضافه کن
  } | null;
  products: UserProductView[];
};

// --------- تایپ‌های استفاده‌شده در UI فعلی (بدون تغییر نام‌) ---------

export type ProductSummary = {
  id?: number;
  brandTitle?: string;
  modelName?: string;
  imageUrl?: string | null;
};

export type UserProductVM = {
  id: number;
  isVisible?: boolean;     // اگر داشتی
  isDollar?: boolean;
  dollarPrice?: string | null;
  otherCosts?: string | null;
  finalPrice?: string | number;
  createdAt?: string;
  product?: ProductSummary;
};

export type PriceListVM = {
  usdPrice?: number | string | null;
};

// --------- Payloads (client -> API) ----------

export type CreateUserProductPayload = {
  categoryId: number;
  brandId: number;
  modelId: number;
  isDollar: boolean;
  dollarPrice?: DecimalString; // decimal string
  otherCosts?: DecimalString;  // decimal string
  finalPrice: DecimalString;   // decimal string
};


export type UpdateUserProductPayload = {
  id: number;
  isDollar?: boolean;
  dollarPrice?: DecimalString;
  otherCosts?: DecimalString;
  finalPrice?: DecimalString;
};

export type DeleteUserProductPayload = { id: number };
export type ChangeStatusPayload   = { userProductId: number };
export type ChangeOrderPayload    = { topProductId: number; bottomProductId: number };
export type FetchByFilterPayload  = { categoryId?: number; searchText?: string };
